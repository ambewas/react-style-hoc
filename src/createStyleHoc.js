import { curry } from 'ramda';

function isStateless(Component) {
  console.log('Component.prototype',Component.prototype);
  return Component.prototype ? Component.prototype.name : false;
}

const generateStyleObject = (style) => {
  // it's a style object with defaults
  if (typeof style === 'object' && !Array.isArray(style)) {
    return style;
  }

  // just an array with style keys
  if (Array.isArray(style)) {
    return Object.assign(...style.map(key => ({ [key]: 30 })));
  }

  // a style key string
  return Object.assign({ [style]: 30 });
};



export const createStyleHoc = curry((style, value, WrappedComponentOrStyle) => {
  // style -> e.g. { margin: 10, padding: 30 } || ['margin', 'padding'] || 'margin'
  // we should support default styles, an array of style values to be added, or one string.

  console.log('value',value);
  let newStyleObject = generateStyleObject(style);
  console.log('newStyleObject', newStyleObject);
  console.log('WrappedComponentOrStyle',WrappedComponentOrStyle);
  // if (!isStateless(WrappedComponentOrStyle)) {
  //   console.log('style, not a component..');
  //   return {
  //     ...WrappedComponentOrStyle,
  //     ...newStyleObject,
  //   };
  // }

  return ({ style, ...props }) => {
    return <WrappedComponentOrStyle style={{ ...newStyleObject, ...style }} { ...props} />;
  };
});
