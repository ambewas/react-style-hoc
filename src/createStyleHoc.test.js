import React from 'react'; // eslint-disable-line
import renderer from 'react-test-renderer';
import { createStyleHoc } from './createStyleHoc.js';

const TestComponent = ({ style }) => <div style={style}>unstyled</div>;

describe('createStyleHoc', () => {


  it('should return a HOC', () => {
    expect(createStyleHoc()).toBeInstanceOf(Function);
    expect(createStyleHoc('margin')).toBeInstanceOf(Function);
    expect(createStyleHoc({margin: 10})).toBeInstanceOf(Function);
    expect(createStyleHoc({margin: 10}, '')).toBeInstanceOf(Function);
  });

  test('when wrapping a component in a HOC created by createStyleHoc, it should add the correct style prop', function() {
    const withMargin = createStyleHoc('padding', 10);
    const Styled = withMargin(TestComponent); // eslint-disable-line

    const component = renderer.create(
      <Styled>qdsf</Styled>
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
    expect(tree.props.style).toEqual({ padding: 10 });
  });

  test('when using createStyleHoc with a style object, it should add the correct style prop', function() {
    const withMargin = createStyleHoc({padding: 10, margin: 40}, '');
    const Styled = withMargin(TestComponent); // eslint-disable-line

    const component = renderer.create(
      <Styled>qdsf</Styled>
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
    expect(tree.props.style).toEqual({ padding: 10, margin: 40 });
  });
});

