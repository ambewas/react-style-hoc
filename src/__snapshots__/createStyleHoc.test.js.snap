// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createStyleHoc when using createStyleHoc with a style object, it should add the correct style prop 1`] = `
<div
  style={
    Object {
      "margin": 40,
      "padding": 10,
    }
  }
>
  unstyled
</div>
`;

exports[`createStyleHoc when wrapping a component in a HOC created by createStyleHoc, it should add the correct style prop 1`] = `
<div
  style={
    Object {
      "padding": 10,
    }
  }
>
  unstyled
</div>
`;
